text/0162-refactoring-support.md

```diff
@@ -0,0 +1,549 @@
+}
+```
+
+If you want to execute only the automatic refactoring, without deploying new
+resources or removing existing ones, you can use the `cdk refactor` command.
+
+### Trigger
+
+This feature will also be available as a [CDK Trigger], which you can enable by
+passing the flag `autoRefactor` to your application:
+
+```typescript
+const app = new App({
+  autoRefactor: true,
+  ...
+});
+```
+
+If this is enabled, the CLI will skip the refactoring step during deployment,
+and a Lambda function will execute it instead, as part of the provisioning
+process by CloudFormation.
+
+---
+
+Ticking the box below indicates that the public API of this RFC has been
+signed-off by the API bar raiser (the `status/api-approved` label was applied to
+the RFC pull request):
+
+```
+
+[ ] Signed-off by API Bar Raiser @rix0rrr
+
+```
+
+## Public FAQ
+
+### What are we launching today?
+
+A new developer experience for CDK users, that allows them to change the
+location of a construct (stack plus logical ID) without causing resource
+replacement. This new experience is available in the CDK CLI `deploy`
+and `refactor` commands, as well as a CDK trigger.
+
+### Why should I use this feature?
+
+If you ever find yourself doing one of the following, you will benefit from
+stack refactoring support:
+
+- Renaming constructs, either intentionally or by mistake.
+- Moving constructs within the same stack. This could be just for better
+  organization, or to create reusable components.
+- Moving constructs between different stacks.
+- Renaming stacks.
+
+### Can the CLI help me resolve ambiguity when refactoring resources?
+
+Not at the moment. One of the constraints we imposed on this feature is that it
+should work in any environment, including CI/CD pipelines, where there is no
+user to answer questions. Although we could easily extend this feature to
+include ambiguity resolution for the interactive case, it wouldn't transfer well
+to the non-interactive case. If you are interested in an in-depth explanation of
+the problem and a possible solution, check Appendix A.
+
+### What if I can't use the CDK CLI in my pipeline?
+
+Some customers have their own mechanisms for deploying stacks to AWS, that don't
+use the CDK CLI. If that is your case, there is still a way you can use this
+feature. The refactoring logic will also be released as a Node.js library and a
+standalone CLI tool (names TBD). If you can incorporate any of these into your
+deployment tooling, you will have the same functionality as the `refactor`
+command in the main CDK CLI. This command uses assumes a specific role, with a
+narrow set of permissions.
+
+Alternatively, you can use this feature via a [trigger](#trigger).
+
+### What if the deployment fails?
+
+After refactoring the stack, the CLI will proceed with the deployment
+(assuming that is your choice). If the deployment fails, and CloudFormation
+rolls it back, the CLI will execute a second refactor, in reverse, to bring the
```

Comment:
if the deployment fails and CFN rolls it back, why does CDK CLI need to execute a second refactor in reverse? is it because the location of resources in tree is already updated in CDK causing a drift in cfn?

Reply:
Because, from the user's point of view, the deployment is atomic: either all of it happens (refactor + deployment), or neither of them. If the deployment fails, CloudFormation will only roll back the deployment. For CloudFormation, the refactoring was an independent operation that happened before and was successful. So, renaming in reverse is the CDK CLI way of completing the rollback, from the user's perspective.